File: c/2015/07/main.c
#include "../../util.h"
#include "../util2015.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif


void part1() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " ");
    int64_t k = 0;
    MAP map;
    parse(&map, in);
    MAP cache;
    cache.values = (RECORD*) malloc(in->length * sizeof(RECORD));
    cache.length = 0;
    printf("Part 1: %ld\n", evaluate(&map, &cache, "a"));
    free(map.values);
    free(cache.values);
    freeArrMulString(in);
}

void part2() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " ");
    int64_t k = 0;
    MAP map;
    parse(&map, in);
    int16_t index = searchMap(&map, "b");
    strcpy(map.values[index].function, "value");
    strcpy(map.values[index].value1, "46065");
    MAP cache;
    cache.values = (RECORD*) malloc(in->length * sizeof(RECORD));
    cache.length = 0;
    printf("Part 2: %ld\n", evaluate(&map, &cache, "a"));
    free(map.values);
    free(cache.values);
    freeArrMulString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}

File: python/2015/07/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u
import util2017 as w

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    return u.read_match(file, "(.+) -> (.+)")

def part1():
    A = read_input()
    k = 0
    d = w.build(A)
    v = dict()
    print("Part 1:",w.execute(d, v,'a'))

def part2():
    A = read_input()
    k = 0
    d = w.build(A)
    d['b'] = ('value', '46065')
    v = dict()
    print("Part 2:",w.execute(d, v,'a'))

part1()
part2()
